llvm/lib/CAS/BuiltinCAS.cpp

```diff
@@ -2687,20 +2687,28 @@ OnDiskCAS::getObjectProxy(IndexProxy I) const {
 
   // Helper for creating the return.
   auto createProxy = [&](MemoryBufferRef Buffer) -> ObjectProxy {
-    assert(Buffer.getBuffer().drop_back(Blob0).end()[0] == 0 &&
-           "Null termination");
-    if (Blob)
+    if (Blob) {
+      assert(Buffer.getBuffer().drop_back(Blob0).end()[0] == 0 &&
+             "Null termination should be guaranteed by creation logic");
       return ObjectProxy{I.Offset, Object, I.Hash, None,
                          toArrayRef(Buffer.getBuffer().drop_back(Blob0))};
-    return ObjectProxy{I.Offset, Object, I.Hash,
-                       DataRecordHandle::get(Buffer.getBuffer().data()), None};
+    }
+
+    DataRecordHandle Record = DataRecordHandle::get(Buffer.getBuffer().data());
+    assert(Record.getData().end()[0] == 0 &&
+           "Null termination should be guaranteed by creation logic");
+    return ObjectProxy{I.Offset, Object, I.Hash, Record, None};
   };
 
   // Check if we've loaded it already.
   if (Optional<MemoryBufferRef> Buffer = StandaloneData.lookup(I.Hash))
     return createProxy(*Buffer);
 
   // Load it from disk.
+  //
+  // Note: Creation logic guarantees that data that needs null-termination is
+  // suitably 0-padded. Requiring null-termination here would be too expensive
+  // for extremely large objects that happen to be page-aligned.
```

Comment:
Note: I added this comment to dissuade future attempts to make the change from https://github.com/apple/llvm-project/pull/4059; if the assertions above fire, then the bug is either:
- The objects were created incorrectly.
- The objects have since been corrupted.
- The assertions are wrong. (The assertions are there to catch the other two cases.)